src/main/java/org/sagebionetworks/bridge/models/schedules2/adherence/weekly/WeeklyAdherenceReportGenerator.java

```diff
@@ -65,46 +81,105 @@ public WeeklyAdherenceReport generate(AdherenceState state) {
+                row.setLabel(displayLabel);
+                row.setSearchableLabel(searchableLabel);
+                row.setSessionGuid(oneDay.getSessionGuid());
+                row.setSessionName(oneDay.getSessionName());
+                row.setSessionSymbol(oneDay.getSessionSymbol());
+                row.setStudyBurstId(oneDay.getStudyBurstId());
+                row.setStudyBurstNum(oneDay.getStudyBurstNum());
+                row.setWeek(oneDay.getWeek());
+                rows.add(row);
             }
         }
-
+        
+        // Now pad the report to fix the number and position of the row entries, so the rows
+        // can be read straight through in any UI of the report. Sort the rows so they display
+        // in an order that makes sense.
+        List<WeeklyAdherenceReportRow> rowList = Lists.newArrayList(rows);
+        rowList.sort(ROW_COMPARATOR);
+        
+        for (int i=0; i < 7; i++) { // day of week
+            List<EventStreamDay> paddedDays = new ArrayList<>();
+            
+            for (WeeklyAdherenceReportRow row : rowList) {
+                List<EventStreamDay> days = finalReport.getByDayEntries().get(i);
+                EventStreamDay oneDay = padEventStreamDay(days, row.getSessionGuid());
+                paddedDays.add(oneDay);
+            }
+            finalReport.getByDayEntries().put(i, paddedDays);
+        }
+        
         // If the report is empty, then we seek ahead and store information on the next activity
-        EventStreamDay nextDay = getNextActivity(state, finalReport, reports);
+        EventStreamDay nextDay = getNextActivity(state, finalReport, eventReport);
         int percentage = AdherenceUtils.calculateAdherencePercentage(ImmutableList.of(finalReport));
-
+        
         WeeklyAdherenceReport report = new WeeklyAdherenceReport();
         report.setByDayEntries(finalReport.getByDayEntries());
         report.setCreatedOn(state.getNow());
         report.setClientTimeZone(state.getClientTimeZone());
         report.setWeeklyAdherencePercent(percentage);
         report.setNextActivity(NextActivity.create(nextDay));
         report.setLabels(labels);
+        report.setRows(rowList);
         
+        // Remove this information from the day entries, as we've moved it to rows in this report.
+        for (List<EventStreamDay> days : report.getByDayEntries().values()) {
+            for (EventStreamDay oneDay : days) {
+                oneDay.setSessionName(null);
+                oneDay.setSessionSymbol(null);
+                oneDay.setStudyBurstId(null);
+                oneDay.setStudyBurstNum(null);
+                oneDay.setWeek(null);
+            }
+        }
         return report;
     }
     
+    private EventStreamDay padEventStreamDay(List<EventStreamDay> days, String sessionGuid) {
+        if (days != null) {
+            Optional<EventStreamDay> oneDay = days.stream()
+                    .filter(day -> day.getSessionGuid().equals(sessionGuid))
```

Comment:
This is a problem because it's not impossible to have multiple triggers of the same session. Find a better way to match.

Reply:
If a session is triggered twice by two different events, it can appear twice in rows. If the events have nearly the same timestamp, the GUIDs and the labels would be exactly the same. Carrying over the session instance GUID from the event stream report (where I've added it) as it is unique between the two streams and will match the correct row to the correct day entry.